grafana/resource_alerting_mute_timing.go

```diff
@@ -0,0 +1,307 @@
+		TimeIntervals: unpackIntervals(intervals),
+	}
+	return mt
+}
+
+func packIntervals(nts []gapi.TimeInterval) []interface{} {
+	if nts == nil {
+		return nil
+	}
+
+	intervals := make([]interface{}, 0)
+	for _, ti := range nts {
+		in := map[string][]interface{}{}
+		if ti.Times != nil {
+			times := []interface{}{}
+			for _, time := range ti.Times {
+				times = append(times, packTimeRange(time))
+			}
+			in["times"] = times
+		}
+		if ti.Weekdays != nil {
+			wkdays := make([]interface{}, 0)
+			for _, wd := range ti.Weekdays {
+				wkdays = append(wkdays, wd)
+			}
+			in["weekdays"] = wkdays
+		}
+		if ti.DaysOfMonth != nil {
+			mdays := make([]interface{}, 0)
+			for _, dom := range ti.DaysOfMonth {
+				mdays = append(mdays, dom)
+			}
+			in["days_of_month"] = mdays
+		}
+		if ti.Months != nil {
+			ms := make([]interface{}, 0)
+			for _, m := range ti.Months {
+				ms = append(ms, m)
+			}
+			in["months"] = ms
+		}
+		if ti.Years != nil {
+			ys := make([]interface{}, 0)
+			for _, y := range ti.Years {
+				ys = append(ys, y)
+			}
+			in["years"] = ys
+		}
+		intervals = append(intervals, in)
+	}
+
+	return intervals
+}
+
+func unpackIntervals(raw []interface{}) []gapi.TimeInterval {
+	if raw == nil {
+		return nil
+	}
+
+	result := make([]gapi.TimeInterval, len(raw))
+	for i, r := range raw {
+		interval := gapi.TimeInterval{}
+		block := r.(map[string]interface{})
+
+		if vals, ok := block["times"]; ok && vals != nil {
+			vals := vals.([]interface{})
+			interval.Times = make([]gapi.TimeRange, len(vals))
+			for i := range vals {
+				interval.Times[i] = unpackTimeRange(vals[i])
+			}
+		}
+		if vals, ok := block["weekdays"]; ok && vals != nil {
+			vals := vals.([]interface{})
+			interval.Weekdays = make([]gapi.WeekdayRange, len(vals))
+			for i := range vals {
+				interval.Weekdays[i] = gapi.WeekdayRange(vals[i].(string))
+			}
+		}
+		if vals, ok := block["days_of_month"]; ok && vals != nil {
+			vals := vals.([]interface{})
+			interval.DaysOfMonth = make([]gapi.DayOfMonthRange, len(vals))
+			for i := range vals {
+				interval.DaysOfMonth[i] = gapi.DayOfMonthRange(vals[i].(string))
+			}
+		}
+		if vals, ok := block["months"]; ok && vals != nil {
+			vals := vals.([]interface{})
+			interval.Months = make([]gapi.MonthRange, len(vals))
+			for i := range vals {
+				interval.Months[i] = gapi.MonthRange(vals[i].(string))
+			}
+		}
+		if vals, ok := block["years"]; ok && vals != nil {
+			vals := vals.([]interface{})
+			interval.Years = make([]gapi.YearRange, len(vals))
+			for i := range vals {
+				interval.Years[i] = gapi.YearRange(vals[i].(string))
+			}
+		}
+
+		result[i] = interval
+	}
+
+	return result
+}
+
+func packTimeRange(time gapi.TimeRange) interface{} {
+	return map[string]string{
+		"start": time.StartMinute,
+		"end":   time.EndMinute,
+	}
+}
+
+func unpackTimeRange(raw interface{}) gapi.TimeRange {
+	vals := raw.(map[string]interface{})
+	return gapi.TimeRange{
+		StartMinute: vals["start"].(string),
```

Comment:
Do we have a guarantee that this will be a string? I assume yes, as the type validation should run prior as part of the runtime.

Reply:
Yes, it's the same type as whatever we packed it into the state as. So the `pack` function directly above this one sort of mirrors it, and packs the other way around from `string` to `interface{}`.

This is pretty common practice in the provider, you'll see lots of type casting among all the types when reading from the data object. It's all over the place in nearly every file.